_SPIKE_.md

```diff
@@ -0,0 +1,297 @@
+
+## Solution details
+
+### Interface describing model links
+
+To describe navigation properties for TypeScript compiler, application
+developers will define a new interface for each model.
+
+BelongsTo relation:
+
+```ts
+@model()
+class Product extends Entity {
+  @belongsTo(() => Category)
+  categoryId: number;
+}
+
+interface ProductLinks {
+  category?: Category & CategoryLinks;
+}
+```
+
+HasMany relation:
+
+```ts
+@model()
+class Category extends Entity {
+  @hasMany(() => Product)
+  products?: Product[];
+}
+
+interface CategoryLinks {
+  products?: Product & ProductLinks;
+}
+```
+
+This solution has few important properties I'd like to explicitly point out:
+
+- It is independent on how the relations are defined. `@belongsTo` decorator is
+  applied on the foreign key property, `@hasMany` decorator is applied to a kind
+  of a navigational property. If we decide to apply relational decorators at
+  class level in the future, this solution will support that too.
+
+- It does not trigger circular-reference bug in design:type metadata, see
+  https://github.com/Microsoft/TypeScript/issues/27519
+
+- It makes it easy to define a type where all navigational properties are
+  optional. For example: `Product & Partial<ProductLinks>`
+
+### Integration with CrudRepository APIs
+
+The CRUD-related Repository interfaces and classes are accepting a new generic
+argument `Links` that's describing navigational properties.
+
+Example use in application-level repositories:
+
+```ts
+export class CategoryRepository extends DefaultCrudRepository<
+  Category,
+  typeof Category.prototype.id,
+  CategoryLinks
+> {
+  // (no changes here)
+}
+```
+
+### OpenAPI Schema generation
+
+When building OpenAPI Schema from a model definition, we provide two modes:
+
+- Own properties only
+- Both own properties and navigational properties
+
+The decision is controlled by a new option passed to `modelToJsonSchema` and
+related methods.
+
+```ts
+// own properties only
+const spec = getJsonSchema(Product);
+// include navigational properties too
+const spec = getJsonSchema(Product, {includeRelations: true});
+```
+
+An example of the produced schema:
+
+```js
+{
+  title: 'CategoryWithLinks',
+  properties: {
+    // own properties
+    id: {type: 'number'},
+    // navigational properties
+    products: {
+      type: 'array',
+      items: {$ref: '#/definitions/ProductWithLinks'},
+    },
+  },
+  definitions: {
+    ProductWithLinks: {
+      title: 'ProductWithLinks',
+      properties: {
+        // own properties
+        id: {type: 'number'},
+        categoryId: {type: 'number'},
+        // navigational properties
+        category: {$ref: '#/definitions/CategoryWithLinks'},
+      },
+    },
+  },
+}
+```
+
+To support integration with OpenAPI spec of controllers, where we want to
+reference a shared definition (component schema), we need a slightly different
+schema. Here is an example as produced by `getJsonSchemaRef`:
```

Comment:
Can you highlight the difference in the produced schema between the examples below and above?